### DISCUSSION TITLE
Propose & Evaluate New Components For Aspire

### DISCUSSION THREAD

**Discussion by josephaw1022** [microsoft/aspire#1570] [Ideas]
I didn't feel like writing out a bunch of individual issues for each and every component, so I am just going to list some ideas out for some components that I think would be nice to haves for Aspire and am looking for feedback on potential ideas, push back on certain ones, which ones seem the easiest or realistic, or whatever is on your mind. Anyway here's what I've came up with.

I didn't feel like writing out a bunch of individual issues for each and every component, so I am just going to list some ideas out for some components that I think would be nice to haves for Aspire and am looking for feedback on potential ideas, push back on certain ones, which ones seem the easiest or realistic, or whatever is on your mind. Anyway here's what I've came up with. 

## Database
- Mariab DB with and without EF Core

## Storage
- Minio (for those on AWS and on prem)

## Secrets / Config
- Vault (for those who can't utilize azure keyvault for whatever the reason)

## Messaging
- Kafka (issue already exists, having it include the Kafka UI image too)
- NATS
- MQTT

## Feature Flags
- Unleash

## Emails
- Mailpit (local email smtp server)

## Data Aggregating
- fluentd
- Logstash
- Debezium

## Api Gateways
- Kong

## Feature Flags
- Unleash

## Emails
- Mailpit (local email smtp server)

## Search
- Elastic Search and Kibana
- Meilisearch
- Solr 

## Container Registries
- Registry2
- Harbor Registry

## CI/CD
- Jenkins
- Gitlab CI/CD workers

## Devtools
- Data API builder for Azure Databases
- Kusto emulator
- Rancher
- Portainer
- Semaphore

## CLIs
- Azure CLI
- AWS CLI
- etc... 

### Note
Some of these I think I would are more suited to be more additions / iterations to what's in place like search and emails.

Some of these I items aren't necessarily tied to a typical 'developer' workflow, but things that I do or have done locally and that I think would be nice to haves like cloud cli and or rancher. Just so that a non dev could easily spin up some dev ops containers and resources and see the relationship between the resources, the same way aspire is doing for .web, api, db, cache, and queue applications right now.

**Comment by davidfowl** [comment]
- [ ] - Mariab DB with and without EF Core
- [x] - NATS https://github.com/dotnet/aspire/pull/1175
- [ ] - Kusto emulator (looks interesting but we don't have a kusto anything so we'd need the service too)
- [ ] - Elastic Search and Kibana (this is interesting to support)

cc @eerhardt

**Reply by eerhardt** [reply]
> Mariab DB with and without EF Core

Is this covered with
* https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.MySqlConnector
* https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql

or is there another Maria DB .NET library?

**Reply by davidfowl** [reply]
Maybe the only work for Maria DB is the resource side of things.

**Comment by EwoutdBoer** [comment]
I would like to vote for Elastic search as well

**Reply by josephaw1022** [reply]
The only challenge I see is the newer versions of Elastic are secure by default and require some manual configuration and tinkering with the containers.

**Comment by g-martin772** [comment]
I would vote for the local mail server (also with an example on how to use it with identity framework to send actual verification emails)

**Reply by davidfowl** [reply]
https://learn.microsoft.com/en-us/dotnet/aspire/extensibility/custom-resources

**Comment by dominiclapointe** [comment]
MQTT would be pretty useful for IOT projects :D

**Reply by mitchdenny** [reply]
I'd happily review a MQTT broker PR if you wanted to put one up :)

**Reply by CZEMacLeod** [reply]
I think it would be pretty easy to wrap the [dotnet/MqttNet](https://github.com/dotnet/MqttNet) broker as an aspire resource.
I hope the next version of that project follows the M.E.* approach for configuration and services a bit more, which would make hooking up service defaults a little easier for consuming via clients.

**Comment by BartNetJS** [comment]
Are azure components like for example:
* Microsoft.Network/trafficManagerProfiles
* Microsoft.ManagedIdentity/userAssignedIdentities
* Microsoft.Network/publicIPAddresses
* Microsoft.Network/applicationGateways
* etc...

Also good candidates for new hosting components?
And what is the best way to get started to contribute,
I've already took a look at https://github.com/dotnet/aspire/blob/main/src/Components/README.md
Any suggestions?

**Reply by eerhardt** [reply]
I believe these are all blocked on waiting for support by the Azure.Provisioning APIs. (but I could be wrong)

cc @mitchdenny @tg-msft

**Reply by davidfowl** [reply]
Correct (we have user assigned identities already), but they won't all be first class in the application model API. You can use the lower level AddAzureConstruct or AddBicepResource to model all of azure. That's not the goal of aspire.

**Comment by amirex128** [comment]
i vote for Elasticsearch😥

**Reply by davidfowl** [reply]
https://github.com/dotnet/aspire/pull/4430

**Comment by sp00ky** [comment]
Some AWS S3-compatible local service

**Reply by josephaw1022** [reply]
> Some AWS S3-compatible local service

https://github.com/dotnet/aspire/issues/1801

**Comment by los93sol** [comment]
I'm interested in Hashicorp Vault and started looking at this recently, but I'm not sure how to proceed.  At this point it starts the container and I've built extension methods for the relevant volumes, but I'd like to run it in a production mode.  That's where I'm stuck because I need some kind of hook to run after the container starts so I can initialize it, configure auth, etc., then persist the relevant information to the service discovery.

**Reply by mitchdenny** [reply]
@los93sol if you would like to have a go at getting an extension for Hashicorp Vault added to the source tree then the best way to get started is create an issue using the API proposal template. Start there so we can get consensus on what the API shape should look like in terms of `Aspire.Hosting.*` and what (if any) `Aspire.*` component package would be required.

Part of this would be figuring out which client library we would want to take a dependency on. VaultSharp looks like it has received more recent updates.

**Reply by los93sol** [reply]
https://github.com/dotnet/aspire/issues/4753

**Reply by KieranFoot** [reply]
I too would like to see Hashicorp Vault be included as an alternative to Azure Key Vault, since currently there is no available alternative.

**Comment by alrz** [comment]
[Hangfire](https://github.com/HangfireIO/Hangfire) with [Hangfire.Redis.StackExchange ](https://github.com/marcoCasamento/Hangfire.Redis.StackExchange) would be a nice addition.

**Reply by mitchdenny** [reply]
@alrz you should be able to connect Hangfire as is to Redis using the connection string that we inject for Redis. Have you tried it?

**Comment by NapalmCodes** [comment]
I wrote one for API Gateway KrakenD:

https://github.com/NapalmCodes/Aspire.Hosting.Krakend

**Comment by davidfowl** [comment]
See the https://github.com/CommunityToolkit/Aspire/issues for these integration proposals!
